enclave-boundary/src/untrusted.rs

```diff
@@ -35,27 +35,45 @@ pub fn make_variable_length_ecall(
         }
 
         let mut outbuf = vec![0u8; outbuf_used];
-        match unsafe {
-            ecall_fcn(
+        unsafe {
+            let result = ecall_fcn(
                 eid,
                 &mut retval,
                 inbuf.as_ptr(),
                 inbuf.len(),
                 outbuf.as_mut_ptr(),
                 outbuf.len(),
                 &mut outbuf_used,
-                &mut outbuf_retry_id,
-            )
-        } {
-            sgx_status_t::SGX_SUCCESS => match retval {
-                sgx_status_t::SGX_ERROR_OUT_OF_MEMORY => continue,
-                sgx_status_t::SGX_SUCCESS => {
-                    outbuf.truncate(outbuf_used);
-                    break Ok(outbuf);
+                &mut outbuf_retry_id
+            );
+            match result {
+                sgx_status_t::SGX_SUCCESS => match retval {
+                    sgx_status_t::SGX_ERROR_OUT_OF_MEMORY => continue,
+                    sgx_status_t::SGX_SUCCESS => {
+                        outbuf.truncate(outbuf_used);
+                        break Ok(outbuf);
+                    },
+                    // Fatal runtime errors
+                    sgx_status_t::SGX_ERROR_INVALID_FUNCTION |
+                        sgx_status_t::SGX_ERROR_OUT_OF_TCS |
+                        sgx_status_t::SGX_ERROR_ENCLAVE_CRASHED |
+                        sgx_status_t::SGX_ERROR_ECALL_NOT_ALLOWED |
+                        sgx_status_t::SGX_ERROR_OCALL_NOT_ALLOWED |
+                        sgx_status_t::SGX_ERROR_STACK_OVERRUN => {
```

Comment:
I don't think it's a blocker but with the number of enums compared and copied between the two return values, suggest adding a local function like
```rust
fn is_fatal_sgx_status(status: sgx_status_t) -> bool {
    match status {
        a | b | c => true,
        _ => false,
    }
}
```

Reply:
OK, I made a commit that does this and moves that check outside of the match blocks. It should preserve the order of the original checks.

Reply:
I think that's broken, you shouldn't be looking at `retval` unless the `result` is successful.
I think the right approach would be to go back to the original match statement and do something like this:
```
        match unsafe {
            ecall_fcn(
                eid,
                &mut retval,
                inbuf.as_ptr(),
                inbuf.len(),
                outbuf.as_mut_ptr(),
                outbuf.len(),
                &mut outbuf_used,
                &mut outbuf_retry_id,
            )
        } {
            sgx_status_t::SGX_SUCCESS => match retval {
                sgx_status_t::SGX_ERROR_OUT_OF_MEMORY => continue,
                sgx_status_t::SGX_SUCCESS => {
                    outbuf.truncate(outbuf_used);
                    break Ok(outbuf);
                }
                other_retval => {
                    if is_fatal_sgx_status(other_retval) {
                        panic!(...);
                    }
                    break (Err(other_retval)),
                }
            },
            status => {
                    if is_fatal_sgx_status(status) {
                        panic!(...);
                    }
                 break Err(status),
            }
 ```